### ISSUE TITLE
Vetur crashes if checkJs is set to true and a vue file which modifies a watched value is opened

### ISSUE THREAD

**Issue by nextgensparx** [vuejs/vetur#751]
- [x] I have searched through existing issues
- [x] I have read through [docs](https://vuejs.github.io/vetur)
- [x] I have read [FAQ](https://github.com/vuejs/vetur/blob/master/docs/FAQ.md)

## Info

- Platform: Linux
- Vetur version: 0.11.7
- VS Code version: 1.21.1

## Problem

When `checkJs` is set to true in `tsconfig.json` or `jsconfig.json` and a `.vue` file which modifies a watched value is opened, Vetur crashes.

![Vetur crash](https://user-images.githubusercontent.com/19307714/38240807-e88eaa00-3730-11e8-84a9-0853a23fad4c.gif)

![vetur crash4](https://user-images.githubusercontent.com/19307714/38242302-bec32800-3734-11e8-8458-b74f640f8c7e.gif)

Vetur crashing on files like this is almost identical to https://github.com/vuejs/vetur/issues/191#issuecomment-300405883

There is no output from Vue Language Server as it appears to crash. However, there is output from the Window log:
```
[renderer3] [error] Channel has been closed: Error: Channel has been closed
	at ExtHostOutputChannel.validate (/usr/share/vscode/resources/app/out/vs/workbench/node/extensionHostProcess.js:35883:23)
	at ExtHostOutputChannel.appendLine (/usr/share/vscode/resources/app/out/vs/workbench/node/extensionHostProcess.js:35863:18)
	at LanguageClient.error (/home/USER/.vscode/extensions/octref.vetur-0.11.7/node_modules/vscode-languageclient/lib/client.js:1592:28)
	at LanguageClient.logFailedRequest (/home/USER/.vscode/extensions/octref.vetur-0.11.7/node_modules/vscode-languageclient/lib/client.js:2050:14)
	at client.sendRequest.then (/home/USER/.vscode/extensions/octref.vetur-0.11.7/node_modules/vscode-languageclient/lib/client.js:721:24)
	at <anonymous>
```

<details>
  <summary>Extensions</summary>
CoenraadS.bracket-pair-colorizer<br>
EditorConfig.EditorConfig<br>
Keno.uikit-3-snippets<br>
Shan.code-settings-sync<br>
Sophisticode.php-formatter<br>
WakaTime.vscode-wakatime<br>
asabil.meson<br>
christian-kohler.npm-intellisense<br>
christian-kohler.path-intellisense<br>
dbaeumer.vscode-eslint<br>
demijollamaxime.bulma<br>
donjayamanne.githistory<br>
ecmel.vscode-html-css<br>
felixfbecker.php-intellisense<br>
formulahendry.auto-close-tag<br>
formulahendry.auto-rename-tag<br>
hnw.vscode-auto-open-markdown-preview<br>
joelday.docthis<br>
mblode.twig-language<br>
minhthai.vscode-todo-parser<br>
ms-python.python<br>
octref.vetur<br>
patrys.vscode-code-outline<br>
rebornix.project-snippets<br>
robertohuertasm.vscode-icons<br>
wayou.vscode-todo-highlight<br>
yycalm.linecount<br>
zhuangtongfa.Material-theme<br>
</details>

## Reproducible Case

Open [nextgensparx/veturpack](https://github.com/nextgensparx/veturpack) in VS Code and then `client/components/Test.vue`.

It's worth noting that if `data2` is changed to `data3` in watch, the issue goes away.

**Comment by HerringtonDarkholme** [comment]
Thanks for reporting. I guess this problem comes from TS server. 

I will investigate more and report issue to related team.

**Comment by HerringtonDarkholme** [comment]
I saw this error:

.vscode/extensions/octref.vetur-0.11.7/server/node_modules/typescript/lib/typescript.js:72587
                throw e;
                ^

TypeError: Cannot read property 'parameters' of undefined
    at inferFromAnnotatedParameters

**Comment by HerringtonDarkholme** [comment]
Boiled down version：

func.ts

```ts
interface ComponentOptions<V> {
  watch: Record<string, WatchHandler<any>>;
}
type WatchHandler<T> = (val: T) => void;
declare function extend(options: ComponentOptions<{}>): void;
export var vextend = extend
```

```js
import {vextend} from './func'
// hover on vextend
export var a = vextend({
  watch: {
    data1(val) {
      this.data2 = 1
    },
    data2(val) { }
  }
})
```
tsconfig.json
```json
{
  "compilerOptions": {
    "checkJs": true,
    "allowJs": true
  }
}
```


Error:

```
Error  - 3:10:23 PM] 'quickinfo' request failed with error.
Error processing request. Cannot set property 'typeParameters' of undefined
TypeError: Cannot set property 'typeParameters' of undefined
    at assignContextualParameterTypes 
```

**Comment by HerringtonDarkholme** [comment]
This issue has already been tracked in https://github.com/Microsoft/TypeScript/issues/23198.

Let's patiently await TS team's fix. Meanwhile we cannot do much in Vue side, so I will close this issue for now.

**Comment by siphomateke** [comment]
@HerringtonDarkholme Can this be fixed now that https://github.com/Microsoft/TypeScript/issues/23198 has been closed? The latest version of Vetur (0.16.2) still crashes on the reproduction [repository](https://github.com/nextgensparx/veturpack) mentioned above.

**Comment by octref** [comment]
![image](https://user-images.githubusercontent.com/4033249/54791085-9592b000-4bf5-11e9-9ede-53303a682a0a.png)

This seems to be solved in TS 3.3, which you can get by:

- `yarn add -D typescript@3.3` in workspace
-  `"vetur.useWorkspaceDependencies": true`

See #682 for details.

This issue will be completely gone after #1163.

**Comment by siphomateke** [comment]
@octref Thanks! That fixes the issue.
